File: src/tao/linux_process_factory.cc
//  File: process_factory.cc
//  Author: Tom Roeder <tmroeder@google.com>
//
//  Description: A factory that creates child processes.
//
//  Copyright (c) 2013, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include "tao/linux_process_factory.h"

#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/resource.h>
#include <wait.h>

#include <algorithm>

#include <glog/logging.h>

#include "tao/pipe_factory.h"
#include "tao/tao_rpc.h"
#include "tao/util.h"

namespace tao {
bool LinuxProcessFactory::MakeHostedProgramSubprin(int id, const string &path,
                                                   string *subprin) const {
  // TODO(kwalsh) Nice toc-tou error here... maybe copy binary to temp file?
  string prog_hash;
  if (!Sha256FileHash(path, &prog_hash)) {
    LOG(ERROR) << "Could not compute the program digest";
    return false;
  }
  subprin->assign(FormatHostedProgramSubprin(id, bytesToHex(prog_hash)));
  return true;
}

bool LinuxProcessFactory::StartHostedProgram(
    const PipeFactory &child_channel_factory, const string &path,
    const list<string> &args, const string &subprin,
    scoped_ptr<HostedLinuxProcess> *child) const {
  scoped_ptr<FDMessageChannel> channel_to_parent, channel_to_child;
  if (!child_channel_factory.CreateChannelPair(&channel_to_parent,
                                               &channel_to_child)) {
    LOG(ERROR) << "Could not create channel for hosted program";
    return false;
  }

  list<int> keep_open;
  if (!channel_to_parent->GetFileDescriptors(&keep_open)) {
    LOG(ERROR) << "Could not get file descriptors for channel to parent";
    exit(1);
    /* never reached */
  }
  keep_open.push_back(STDIN_FILENO);
  keep_open.push_back(STDOUT_FILENO);
  keep_open.push_back(STDERR_FILENO);

  scoped_ptr<TaoRPC> rpc_to_parent(new TaoRPC(channel_to_parent.release()));
  string child_rpc_params;
  if (!rpc_to_parent->SerializeToString(&child_rpc_params)) {
    LOG(ERROR) << "Could not encode child RPC channel parameters";
    return false;
  }
  VLOG(0) << "Channel to parent is: " << child_rpc_params;

  // There is a race between fork-kill and fork-exec. We currently have various
  // signal handlers installed (e.g. SIGTERM for graceful shutdown in Listen
  // loop, SIGPIPE ignored, SIGCHLD for catching child exits). These are
  // inherited across fork. If someone calls StopHostedProgram() very soon after
  // fork, then the kill will happen before the exec, meaning the child will
  // accidentally (mis)handle the SIGTERM. Solution: either use vfork and rely
  // its obscure "block until child execs" semantics, or block signals before
  // the fork then have parent re-enable them and child clear them to defaults.

  // Block all signals.
  sigset_t old_signals, all_signals, no_signals;
  sigfillset(&all_signals);
  sigemptyset(&no_signals);
  if (pthread_sigmask(SIG_SETMASK, &all_signals, &old_signals) < 0) {
    LOG(ERROR) << "Could not block signals";
    return false;
  }

  int child_pid = fork();
  if (child_pid == -1) {
    pthread_sigmask(SIG_SETMASK, &old_signals, nullptr);
    LOG(ERROR) << "Could not fork hosted program";
    return false;
  }

  if (child_pid == 0) {
    // Child

    // Become process group leader.
    setpgrp();

    // Reset all signals
    for (int signum = 1; signum <= NSIG; signum++) {
      struct sigaction sig_act;
      memset(&sig_act, 0, sizeof(sig_act));
      sig_act.sa_handler = SIG_DFL;
      sig_act.sa_flags = 0;
      sigaction(signum, &sig_act, nullptr);
    }

    // Unblock all signals.
    if (pthread_sigmask(SIG_SETMASK, &no_signals, nullptr) < 0) {
      LOG(ERROR) << "Could not unblock signals";
      exit(1);
    }

    int argc = 1 + (int)args.size();     // 1+ for path at start
    char **argv = new char *[argc + 1];  // +1 for null at end
    int i = 0;
    argv[i++] = strdup(path.c_str());
    for (const string &arg : args) {
      argv[i++] = strdup(arg.c_str());
    }
    argv[i++] = nullptr;
    // We couuld put channel params in argv:
    // argv[..] = strdup(child_rpc_params.c_str());
    // Instead, put it in environment variable so we can host Tao-oblivious
    // programs without messing up their argv...
    setenv(Tao::HostTaoEnvVar, child_rpc_params.c_str(), 1);

    channel_to_child->Close();

    close(STDIN_FILENO);
    dup2(open("/dev/null", O_RDONLY), STDIN_FILENO);
    if (!CloseAllFileDescriptorsExcept(keep_open)) {
      LOG(ERROR) << "Could not clean up file descriptors";
      exit(1);
      /* never reached */
    }

    int rv = execv(path.c_str(), argv);
    if (rv == -1) {
      PLOG(ERROR) << "Could not exec " << path;
      exit(1);
    }
    /* never reached */
    CHECK(false);
    return false;
  } else {
    // Parent

    // Unblock old signals.
    if (pthread_sigmask(SIG_SETMASK, &old_signals, nullptr) < 0) {
      LOG(ERROR) << "Could not unblock signals";
      exit(1);
    }

    rpc_to_parent->Close();

    child->reset(new HostedLinuxProcess);
    (*child)->subprin = subprin;
    (*child)->pid = child_pid;
    (*child)->rpc_channel.reset(channel_to_child.release());
    LOG(INFO) << "Started PID " << child_pid << " for " << elideString(subprin);
    return true;
  }
}

bool LinuxProcessFactory::StopHostedProgram(HostedLinuxProcess *child,
                                            int signum) const {
  if (child->pid <= 0) return false;  // already dead or invalid PID
  // There is a race between fork-kill and fork-setpgrp. After fork, if we try
  // to kill the entire child group, the child may not yet have called setpgrp.
  // So we first kill the child individually. If it has not called setpgrp, then
  // we are done. Otherwise there may still be descendents that need to be
  // killed. So in both cases, we next kill the group. If there are no
  // descendents, then no harm is done.
  //
  // Note: There is no race between exit and kill, since PIDs will not be reused
  // before we do waitpid, which zeros out the child pid, or before all
  // group members have exited.
  bool success = true;
  if (kill(child->pid, signum) < 0) {
    PLOG(ERROR) << "Could not stop hosted program with PID " << child->pid;
    success = false;
  } else {
    LOG(INFO) << "Sent signal " << signum << " to hosted program with PID "
              << child->pid;
  }
  if (kill(-1 * child->pid, signum) < 0) {
    LOG(INFO) << "No group members killed for PID " << child->pid;
  } else {
    LOG(INFO) << "Group members also killed for PID " << child->pid;
  }
  return success;
}

int LinuxProcessFactory::WaitForHostedProgram() const {
  int status;
  int pid = waitpid(-1, &status, WNOHANG);
  if (pid == 0) {
    // There are children, but they haven't exited.
  } else if (pid == -1) {
    // There are no children.
  } else if (WIFEXITED(status)) {
    LOG(INFO) << "Hosted process with PID " << pid << " has exited";
  } else if (WIFSIGNALED(status)) {
    LOG(INFO) << "Hosted process with PID " << pid << " has been killed";
  } else {
    LOG(INFO) << "Hosted process with PID " << pid
              << " died for unknown reasons";
  }
  return pid;
}

// TODO(kwalsh) Replace this with formula formatting routines
string LinuxProcessFactory::FormatHostedProgramSubprin(
    int id, const string &prog_hash) const {
  stringstream out;
  if (id != 0)
    out << "Process(" << id << ", ";
  else
    out << "Program(";
  out << quotedString(prog_hash) << ")";
  return out.str();
}

// TODO(kwalsh) Replace this with formula parsing routines
bool LinuxProcessFactory::ParseHostedProgramSubprin(const string &subprin,
                                                    int *id, string *prog_hash,
                                                    string *extension) const {
  stringstream in(subprin);
  if (subprin.substr(0, 8) == "Program(") {
    skip(in, "Program(");
    *id = 0;
    getQuotedString(in, prog_hash);
    skip(in, ")");
  } else {
    skip(in, "Process(");
    in >> *id;
    skip(in, ", ");
    getQuotedString(in, prog_hash);
    skip(in, ")");
  }

  string rest;
  if (in && getline(in, rest, '\0') && rest != "") {
    stringstream in_rest(rest);
    skip(in_rest, "::");
    string ext;
    getline(in_rest, ext, '\0');
    extension->assign(ext);
  } else {
    extension->assign("");
  }

  if (in.bad()) {
    LOG(ERROR) << "Could not parse hosted program subprincipal: " << subprin;
    return false;
  }

  return true;
}

static bool CloseExcept(int fd, const list<int> &keep_open) {
  if (std::find(keep_open.begin(), keep_open.end(), fd) != keep_open.end()) {
    return true;
  } else if (close(fd) < 0 && errno != EBADF) {
    PLOG(ERROR) << "Could not close fd " << fd;
    return false;
  } else {
    return true;
  }
}

bool LinuxProcessFactory::CloseAllFileDescriptorsExcept(
    const list<int> &keep_open) {
  DIR *dir = opendir("/proc/self/fd");
  int dir_fd = dirfd(dir);
  if (dir != nullptr) {
    struct dirent *entry;
    while ((entry = readdir(dir)) != nullptr) {
      if (entry->d_name[0] == '.') continue;
      char *end = nullptr;
      errno = 0;
      int n = strtol(entry->d_name, &end, 10);
      if (errno != 0) {
        PLOG(ERROR) << "Error enumerating /proc/self/fd";
        closedir(dir);
        return false;
      } else if (end == nullptr || end == entry->d_name || end[0] != '\0' ||
                 n < 0 || n > INT_MAX) {
        LOG(ERROR) << "Error enumerating /proc/self/fd";
        closedir(dir);
        return false;
      }
      int fd = static_cast<int>(n);
      if (fd != dir_fd && !CloseExcept(fd, keep_open)) {
        closedir(dir);
        return false;
      }
    }
    closedir(dir);
    return true;
  } else {
    struct rlimit limits;
    if (getrlimit(RLIMIT_NOFILE, &limits) < 0) {
      LOG(ERROR) << "Could not get rlimits";
      return false;
    }
    for (int fd = 0; fd < static_cast<int>(limits.rlim_max); fd++) {
      if (fd != dir_fd && !CloseExcept(fd, keep_open)) {
        closedir(dir);
        return false;
      }
    }
    return true;
  }
}

}  // namespace tao

File: go/tao/linux_process_factory.go
// Copyright (c) 2014, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tao

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"os/signal"
	"path"
	"strings"
	"syscall"

	"github.com/jlmucb/cloudproxy/go/tao/auth"
	"github.com/jlmucb/cloudproxy/go/util"
)

// A LinuxProcessFactory supports methods for creating Linux processes as
// hosted programs. LinuxProcessFactory implements HostedProgramFactory.
type LinuxProcessFactory struct {
	channelType string
	socketPath  string
}

// NewLinuxProcessFactory returns a new HostedProgramFactory that can create
// linux processes.
func NewLinuxProcessFactory(channelType, socketPath string) HostedProgramFactory {
	return &LinuxProcessFactory{
		channelType: channelType,
		socketPath:  socketPath,
	}
}

// A LinuxProcess represents a hosted program that executes as a linux process.
type HostedProcess struct {

	// The spec from which this process was created.
	spec HostedProgramSpec

	// The value to be used as argv[0]
	Argv0 string

	// A secured, private copy of the executable.
	Temppath string

	// A temporary directory for storing the temporary executable.
	Tempdir string

	// Hash of the executable.
	Hash []byte

	// The underlying process.
	Cmd exec.Cmd

	// The factory responsible for the hosted process.
	Factory *LinuxProcessFactory

	// A channel to be signaled when the process is done.
	Done chan bool
}

// NewHostedProgram initializes, but does not start, a hosted process.
func (lpf *LinuxProcessFactory) NewHostedProgram(spec HostedProgramSpec) (child HostedProgram, err error) {

	// The argv[0] for the child is given by spec.ContainerArgs
	argv0 := spec.Path
	if len(spec.ContainerArgs) == 1 {
		argv0 = spec.ContainerArgs[0]
	} else if len(spec.ContainerArgs) > 0 {
		err = fmt.Errorf("Too many container arguments for process")
		return
	}

	// To avoid a time-of-check-to-time-of-use error, we copy the file
	// bytes to a temp file as we read them. This temp-file path is
	// returned so it can be used to start the program.
	tempdir, err := ioutil.TempDir("/tmp", "cloudproxy_linux_host")
	if err != nil {
		return
	}
	defer func() {
		if err != nil {
			os.RemoveAll(tempdir)
		}
	}()
	if err = os.Chmod(tempdir, 0755); err != nil {
		return
	}

	temppath := path.Join(tempdir, "hosted_program")
	tf, err := os.OpenFile(temppath, os.O_CREATE|os.O_RDWR, 0700)
	defer tf.Close()
	if err != nil {
		return
	}
	if err = tf.Chmod(0755); err != nil {
		return
	}

	inf, err := os.Open(spec.Path)
	defer inf.Close()
	if err != nil {
		return
	}

	// Read from the input file and write to the temp file.
	tr := io.TeeReader(inf, tf)
	b, err := ioutil.ReadAll(tr)
	if err != nil {
		return
	}

	h := sha256.Sum256(b)

	child = &HostedProcess{
		spec:     spec,
		Argv0:    argv0,
		Temppath: temppath,
		Tempdir:  tempdir,
		Hash:     h[:],
		Factory:  lpf,
		Done:     make(chan bool, 1),
	}
	return
}

// Use 24 bytes for the socket name.
const sockNameLen = 24

// Start starts the the hosted process and returns a tao channel to it.
func (p *HostedProcess) Start() (channel io.ReadWriteCloser, err error) {
	var extraFiles []*os.File
	var evar string
	switch p.Factory.channelType {
	case "pipe":
		// Get a pipe pair for communication with the child.
		var serverRead, clientRead, serverWrite, clientWrite *os.File
		serverRead, clientWrite, err = os.Pipe()
		if err != nil {
			return
		}
		defer clientWrite.Close()

		clientRead, serverWrite, err = os.Pipe()
		if err != nil {
			serverRead.Close()
			return
		}
		defer clientRead.Close()

		channel = util.NewPairReadWriteCloser(serverRead, serverWrite)
		extraFiles = []*os.File{clientRead, clientWrite} // fd 3, fd 4

		// Note: ExtraFiles below ensures readfd=3, writefd=4 in child
		evar = HostSpecEnvVar + "=tao::RPC+tao::FDMessageChannel(3, 4)"
	case "unix":
		// Get a random name for the socket.
		nameBytes := make([]byte, sockNameLen)
		if _, err = rand.Read(nameBytes); err != nil {
			return
		}
		sockName := base64.URLEncoding.EncodeToString(nameBytes)
		sockPath := path.Join(p.Factory.socketPath, sockName)
		channel = util.NewUnixSingleReadWriteCloser(sockPath)
		if channel == nil {
			err = fmt.Errorf("Couldn't create a new Unix channel\n")
			return
		}
		evar = HostSpecEnvVar + "=" + sockPath
	default:
		err = fmt.Errorf("invalid channel type '%s'\n", p.Factory.channelType)
		return
	}
	defer func() {
		if err != nil {
			channel.Close()
			channel = nil
		}
	}()

	stdin, stdout, stderr, moreFiles := util.NewStdio(p.spec.Files)
	extraFiles = append(extraFiles, moreFiles...)

	env := p.spec.Env
	if env == nil {
		env = os.Environ()
	}
	// Make sure that the child knows to use the right kind of channel.
	etvar := HostChannelTypeEnvVar + "=" + p.Factory.channelType
	replaced := false
	replacedType := false
	for i, pair := range env {
		if strings.HasPrefix(pair, HostSpecEnvVar+"=") {
			env[i] = evar
			replaced = true
		}

		if strings.HasPrefix(pair, HostChannelTypeEnvVar+"=") {
			env[i] = etvar
			replacedType = true
		}
	}
	if !replaced {
		env = append(env, evar)
	}

	if !replacedType {
		env = append(env, etvar)
	}

	if (p.spec.Uid == 0 || p.spec.Gid == 0) && !p.spec.Superuser {
		err = fmt.Errorf("Uid and Gid must be nonzero unless Superuser is set\n")
		return
	}

	wd := p.spec.Dir
	if wd == "" {
		wd = p.Tempdir
	}

	// Every hosted process is given its own process group (Setpgid=true). This
	// ensures that hosted processes will not be in orphaned process groups,
	// allowing them to receive job control signals (SIGTTIN, SIGTTOU, and
	// SIGTSTP).
	//
	// If this host is running in "daemon" mode, i.e. without a controlling tty
	// and in our own session and process group, then this host will be (a) the
	// parent of a process in the child's group, (b) in the same session, and
	// (c) not in the same group as the child, so it will serve as the anchor
	// that keeps the child process groups from being considered orphaned.
	//
	// If this host is running in "foreground" mode, i.e. with a controlling tty
	// and as part of our parent process's session but in our own process group,
	// then the same three conditions are satisified, so this host can still
	// serve as the anchor that keeps the child process groups from being
	// considered orphaned. (Note: We could also use Setpid=false in this case,
	// since the host would be part of the child process group and our parent
	// would then meet the requirements.)

	spa := &syscall.SysProcAttr{
		Credential: &syscall.Credential{
			Uid: uint32(p.spec.Uid),
			Gid: uint32(p.spec.Uid),
		},
		// Setsid: true, // Create session.
		Setpgid: true, // Set process group ID to new pid (SYSV setpgrp)
		// Setctty: true, // Set controlling terminal to fd Ctty (only meaningful if Setsid is set)
		// Noctty: true, // Detach fd 0 from controlling terminal
		// Ctty: 0, // Controlling TTY fd (Linux only)
	}
	argv := []string{p.Argv0}
	argv = append(argv, p.spec.Args...)
	p.Cmd = exec.Cmd{
		Path:        p.Temppath,
		Dir:         wd,
		Args:        argv,
		Stdin:       stdin,
		Stdout:      stdout,
		Stderr:      stderr,
		Env:         env,
		ExtraFiles:  extraFiles,
		SysProcAttr: spa,
	}

	if err = p.Cmd.Start(); err != nil {
		return
	}

	// Reap the child when the process dies.
	sc := make(chan os.Signal, 1)
	signal.Notify(sc, syscall.SIGCHLD)
	go func() {
		<-sc
		p.Cmd.Wait()
		signal.Stop(sc)
		os.RemoveAll(p.Tempdir)
		p.Done <- true
		close(p.Done) // prevent any more blocking
	}()

	// TODO(kwalsh) put channel into p, remove the struct in linux_host.go

	return
}

// ExitStatus returns an exit code for the process.
func (p *HostedProcess) ExitStatus() (int, error) {
	s := p.Cmd.ProcessState
	if s == nil {
		return -1, fmt.Errorf("Child has not exited")
	}
	if code, ok := (*s).Sys().(syscall.WaitStatus); ok {
		return int(code), nil
	}
	return -1, fmt.Errorf("Couldn't get exit status\n")
}

// WaitChan returns a chan that will be signaled when the hosted process is
// done.
func (p *HostedProcess) WaitChan() <-chan bool {
	return p.Done
}

// Kill kills an os/exec.Cmd process.
func (p *HostedProcess) Kill() error {
	return p.Cmd.Process.Kill()
}

// Stop tries to send SIGTERM to a process.
func (p *HostedProcess) Stop() error {
	err := syscall.Kill(p.Cmd.Process.Pid, syscall.SIGTERM)
	syscall.Kill(p.Cmd.Process.Pid, syscall.SIGCONT)
	return err
}

// Spec returns the specification used to start the hosted process.
func (p *HostedProcess) Spec() HostedProgramSpec {
	return p.spec
}

// Pid returns the pid of the underlying os/exec.Cmd instance.
func (p *HostedProcess) Pid() int {
	return p.Cmd.Process.Pid
}

// Subprin returns the subprincipal representing the hosted process.
func (p *HostedProcess) Subprin() auth.SubPrin {
	return FormatProcessSubprin(p.spec.Id, p.Hash)
}

// FormatProcessSubprin produces a string that represents a subprincipal with
// the given ID and hash.
func FormatProcessSubprin(id uint, hash []byte) auth.SubPrin {
	var args []auth.Term
	if id != 0 {
		args = append(args, auth.Int(id))
	}
	args = append(args, auth.Bytes(hash))
	return auth.SubPrin{auth.PrinExt{Name: "Program", Arg: args}}
}

func (p *HostedProcess) Cleanup() error {
	// TODO(kwalsh) close channel, maybe also kill process if still running?
	os.RemoveAll(p.Tempdir)
	return nil
}
